### PR TITLE
fix: add notifications for the correct new accounts

### PR DESCRIPTION
**Pull request by Prithpal-Sooriya**
## **Description**

There was a weird bug, where when the service worker restarts (e.g. browser close/open), we auto subscribed all accounts. Even accounts that a user may have chosen to disable are now re-enabled.

This adds some controller state to track accounts seen. So we can correctly add or remove notifications on the correct accounts.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24758?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts [hunk 1]
```diff
@@ -90,14 +90,16 @@ describe('metamask-notifications - constructor()', () => {
     expect(mockDelete).not.toBeCalled();
   });
 
-  test('Keying Change Event with new triggers', async () => {
+  test('Keying Change Event with new triggers will update triggers correctly', async () => {
```
Comment (danjm) [inline app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:93]: Keying -> Keyring

File: app/scripts/controllers/metamask-notifications/metamask-notifications.ts [hunk 2]
```diff
@@ -281,30 +293,49 @@ export class MetamaskNotificationsController extends BaseController<
...
-
-      const accountsRemoved = [...this.#prevAccountsSet.values()].filter(
+      // Calculate added and removed addresses
+      const accountsAdded = accounts.filter((a) => !prevAccountsSet.has(a));
+      const accountsRemoved = [...prevAccountsSet.values()].filter(
         (a) => !currentAccountsSet.has(a),
       );
 
-      this.#prevAccountsSet = new Set(accounts);
+      console.log({
```
Comment (danjm) [inline app/scripts/controllers/metamask-notifications/metamask-notifications.ts:327]: console.log should be removed

**Review by danjm — COMMENTED** [review]
looks good to me. Remove the console.log and I can approve

File: app/scripts/controllers/metamask-notifications/metamask-notifications.ts (hunk 2 above)
Comment (Prithpal-Sooriya) [reply app/scripts/controllers/metamask-notifications/metamask-notifications.ts:327]: Also good spot, this was a late PR so can fix CI and ask for re-review

**Review by matteoscurati — APPROVED** [review]

**Review by danjm — APPROVED** [review]

**Pull request merged**